packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Models/Types/ModelTypeProvider.cs

```diff
@@ -48,21 +49,41 @@ protected override PropertyDeclaration[] BuildProperties()
         private PropertyDeclaration BuildPropertyDeclaration(InputModelProperty property)
         {
             var propertyType = CodeModelPlugin.Instance.TypeFactory.CreateCSharpType(property.Type);
+            var propHasSetter = PropertyHasSetter(propertyType, property);
             MethodSignatureModifiers setterModifier = !property.IsReadOnly ? MethodSignatureModifiers.Public : MethodSignatureModifiers.None;
 
-            // Exclude setter generation for collection properties https://github.com/Azure/autorest.csharp/issues/4616
             // Add Remarks and Example for BinaryData Properties https://github.com/Azure/autorest.csharp/issues/4617
             var propertyDeclaration = new PropertyDeclaration(
                     description: FormattableStringHelpers.FromString(property.Description),
                     modifiers: MethodSignatureModifiers.Public,
                     propertyType: propertyType,
                     name: property.Name.FirstCharToUpperCase(),
-                    propertyBody: new AutoPropertyBody(!property.IsReadOnly, setterModifier, GetPropertyInitializationValue(property, propertyType))
+                    propertyBody: new AutoPropertyBody(propHasSetter, setterModifier, GetPropertyInitializationValue(property, propertyType))
                     );
 
             return propertyDeclaration;
         }
 
+        /// <summary>
+        /// Returns true if the property has a setter. Read-only properties and non-nullable collections do not have setters.
+        /// </summary>
+        /// <param name="type">The <see cref="CSharpType"/> of the property.</param>
+        /// <param name="prop">The <see cref="InputModelProperty"/>.</param>
+        private bool PropertyHasSetter(CSharpType type, InputModelProperty prop)
+        {
+            if (prop.IsReadOnly)
+            {
+                return false;
+            }
+
+            if (type.IsCollection)
+            {
+                return type.IsNullable;
+            }
+
+            return true;
+        }
```

Comment:
In autorest.csharp, we have all these conditions in [this file](https://github.com/Azure/autorest.csharp/blob/feature/v3/src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProviderFields.cs#L195-L232) that determine whether we need a setter for a property or not.

Particularly for collections, we have [this](https://github.com/Azure/autorest.csharp/blob/feature/v3/src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProviderFields.cs#L224C17-L224C60) condition: `type.IsCollection && !type.IsReadOnlyMemory`. I believe we need to include the condition `type` is not `ReadOnlyMemory` here, especially in the case for vector properties.

Reply:
thanks. I've pulled in some of these additional conditions. Do let me know if you still think the conditions were adding aren't enough to cover all the properties we need to generate.